framework/src/ble_manager/ble_queue.c

```diff
@@ -0,0 +1,198 @@
+/****************************************************************************
+ *
+ * Copyright 2021 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <errno.h>
+#include <ble_manager/ble_queue.h>
+
+#define BLE_QUEUE_CHECK do { if (q_grp == NULL) {return BLE_QUEUE_NOT_INIT;}} while(0)
+
+static ble_queue_group *q_grp = NULL;
+static int g_run_process = 0;
+
+static int _ble_evt_handler(int argc, char *argv[])
+{
+	int priority;
+	int err_no;
+	ble_queue *q;
+
+	while (g_run_process) {
+		if (sem_wait(&q_grp->countsem) < 0) {
+			if (errno == EINTR) {
+				continue;
+			} else {
+				g_run_process = 0;
+				return -1;
+			}
+		}
+		for (priority = 0; priority < BLE_QUEUE_EVT_PRI_MAX; priority++) {
+			q = &q_grp->q[priority];
+			if (sem_trywait(&q->countsem) < 0) {
+				err_no = get_errno();
+				if (err_no == EDEADLK || err_no == EINVAL) {
+					g_run_process = 0;
+					return -1;
+				}
+			} else {
+				if (q->read_index != q->write_index) {
+					/* This Priority Queue is not Empty */
+					q_grp->caller(priority, q->queue + (q->data_size * q->read_index));
+					q->read_index = (q->read_index + 1) % q->size;
+				}
+				break;
+			}
+		}
+	}
+
+	return 0;
+}
+
+ble_queue_ret_e ble_queue_init(queue_event_caller caller)
+{
+	if (q_grp == NULL) {
+		q_grp = (ble_queue_group *)malloc(sizeof(ble_queue_group));
```

Comment:
How about using `zalloc` instead of `malloc`?
Then we don't need to call `memset` from line 71.

Reply:
Okay, I will fix it